File: VB/XtraReport1.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.XtraReports.UI
' ...

Namespace PreviewSort
	Partial Public Class XtraReport1
		Inherits DevExpress.XtraReports.UI.XtraReport
		Public Sub New()
			InitializeComponent()
			currentSortCell = xrTableCell1
		End Sub

		Private currentSortCell As XRTableCell

		Public Sub SetReportParameter(ByVal orderID As Integer)
			nwindDataSet1.Clear()
			orderDetailsTableAdapter.Fill(nwindDataSet1.OrderDetails, orderID)
			lbOrderID.Text = orderID.ToString()
		End Sub

		Private Sub OnCellPreviewClick(ByVal sender As Object, ByVal e As PreviewMouseEventArgs) Handles xrTableCell1.PreviewClick, xrTableCell4.PreviewClick, xrTableCell7.PreviewClick, xrTableCell2.PreviewClick, xrTableCell13.PreviewClick
			' Turn off sorting.
			Me.Detail.SortFields.Clear()
			currentSortCell.Text = currentSortCell.Text.Remove(currentSortCell.Text.Length - 1, 1)

			' Create a new field to sort.
			Dim grField As New GroupField()
			grField.FieldName = (CType(sender, XRControl)).Tag.ToString()
			grField.SortOrder = XRColumnSortOrder.Ascending

			' Add sorting.
			Me.Detail.SortFields.Add(grField)
			CType(sender, XRLabel).Text = (CType(sender, XRLabel)).Text & "*"
			currentSortCell = CType(sender, XRTableCell)

			' Recreate the report document.
			Me.CreateDocument()
		End Sub

	End Class
End Namespace

File: CS/XtraReport1.cs
using System;
using DevExpress.XtraReports.UI;
// ...

namespace PreviewSort {
    public partial class XtraReport1 : DevExpress.XtraReports.UI.XtraReport {
        public XtraReport1() {
            InitializeComponent();
            currentSortCell = xrTableCell1;
        }

        private XRTableCell currentSortCell;

        public void SetReportParameter(int orderID) {
            nwindDataSet1.Clear();
            orderDetailsTableAdapter.Fill(nwindDataSet1.OrderDetails, orderID);
            lbOrderID.Text = orderID.ToString();
        }

        private void OnCellPreviewClick(object sender, PreviewMouseEventArgs e) {
            // Turn off sorting.
            this.Detail.SortFields.Clear();
            currentSortCell.Text = currentSortCell.Text.Remove(currentSortCell.Text.Length - 1, 1);

            // Create a new field to sort.
            GroupField grField = new GroupField();
            grField.FieldName = ((XRControl)sender).Tag.ToString();
            grField.SortOrder = XRColumnSortOrder.Ascending;

            // Add sorting.
            this.Detail.SortFields.Add(grField);
            ((XRLabel)sender).Text = ((XRLabel)sender).Text + "*";
            currentSortCell = (XRTableCell)sender;

            // Recreate the report document.
            this.CreateDocument();
        }

    }
}
